python/triton/language/semantic.py

```diff
@@ -1394,7 +1394,8 @@ def dot(lhs: tl.tensor, rhs: tl.tensor, acc: tl.tensor, input_precision: Optiona
 
 
 def where(condition: tl.tensor, x: tl.tensor, y: tl.tensor, builder: ir.builder) -> tl.tensor:
-    condition = cast(condition, tl.int1, builder)
+    if condition.dtype != tl.int1:
+        raise ValueError(f"tl.where: condition must be a boolean. Got {condition.dtype}")
```

Comment:
This seems reasonable eventually, but unless it's changed recently there are several places where booleans end up with `tl.int8` dtype so I actually expect this would break a lot of code.

Reply:
> there are several places where booleans end up with tl.int8 dtype
Oh? In what cases does that happen?

Reply:
The main example I had in mind is `tl.load`. 

https://github.com/triton-lang/triton/blob/cf31278859a170eb2f20407c9f4004cafdf41c0d/python/triton/language/semantic.py#L997-L1001

There is also e.g. `tl.extra.cuda.libdevice.isinf` which returns `int32`:
https://github.com/triton-lang/triton/blob/cf31278859a170eb2f20407c9f4004cafdf41c0d/python/triton/language/extra/cuda/libdevice.py#L1084-L1087

Reply:
If you look at inductor-generated code I basically added `to(tl.int1)` to as many places as possible to work around these issues.

Reply:
those look more like bugs than features. I think we should just cast those to `tl.int1` at a semantics level. WDYT?

Reply:
@peterbell10 Good point about `tl.load`! That said, I don't imagine that PyTorch boolean arrays are very commonly used in custom kernels: if you're writing a custom Triton kernel instead of using PyTorch then it's probably because you care about efficiency, and if you care about efficiency you'd use a memory-efficient format (e.g. a uint32 array with 32 booleans packed into each element) instead of one that wastes 87.5% of the space and gmem bandwidth.

@lezcano I agree that Triton operations that return booleans, such as that `isinf`, should be `tl.int1`.